dpc/src/state/builder.rs

```diff
@@ -0,0 +1,371 @@
+// Copyright (C) 2019-2021 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::prelude::*;
+use snarkvm_utilities::ToBytes;
+
+use anyhow::{anyhow, Result};
+use rand::{CryptoRng, Rng};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub struct StateBuilder<C: Parameters> {
+    /// A list of given inputs for a state transition.
+    inputs: Vec<Input<C>>,
+    /// A list of expected outputs for a state transition.
+    outputs: Vec<Output<C>>,
+    /// A publicly-visible field with data from the state transition.
+    memo: Vec<u8>,
+    /// A list of errors accumulated from calling the builder.
+    errors: Vec<String>,
+}
+
+impl<C: Parameters> StateBuilder<C> {
+    ///
+    /// Initializes a new instance of `StateBuilder`.
+    ///
+    pub fn new() -> Self {
+        Self {
+            inputs: Vec::new(),
+            outputs: Vec::new(),
+            memo: Vec::new(),
+            errors: Vec::new(),
+        }
+    }
+
+    ///
+    /// Adds the given input into the builder.
+    ///
+    pub fn add_input(mut self, input: Input<C>) -> Self {
+        // Ensure there are no outputs assigned yet, as the builder computes joint serial numbers.
+        if !self.outputs.is_empty() {
+            self.errors.push("Builder cannot add new inputs after outputs".into());
+        }
+
+        match self.inputs.len() < C::NUM_INPUT_RECORDS {
```

Comment:
Can the number of inputs "==" to `C::NUM_INPUT_RECORDS`?

Reply:
It cannot equal, that would exceed then

Reply:
Can you provide more info? Since intuitively it looks like a user can provide all the input records (i.e., no noop padding)?

Reply:
Is there a compulsory padding or?

Reply:
This `<` just enforces there is at least 1 more input slot available